iucvterm/doc/iucvtty.1
.\" iucvtty.1
.\"
.\"
.\" Copyright IBM Corp. 2008, 2017
.\" s390-tools is free software; you can redistribute it and/or modify
.\" it under the terms of the MIT license. See LICENSE for details.
.\" ----------------------------------------------------------------------
.TH iucvtty "1" "March 2009" "s390-tools" "IUCV terminal applications"
.
.ds t \fBiucvtty\fP
.ds i \fBiucvconn\fP
.
.
.
.SH NAME
iucvtty \- allow remote logins over z/VM IUCV
.
.
.
.SH SYNOPSIS
.B iucvtty
.RB [ \-h | \-\-help ]
.br
.B iucvtty
.RB [ \-v | \-\-version ]
.br
.B iucvtty
.RB [ \-a | \-\-allow-from
.IR regex ]
.IR terminal_id
.RB [\-\-
.IR login_program " [" login_options ]]
.
.
.
.SH DESCRIPTION
The \*t application provides full-screen terminal access to a Linux instance
running as a z/VM guest operating system.
The \*i application is used to connect to a running \*t instance.

The terminal connection is based on the z/VM Inter-User Communication Vehicle
(IUCV). z/VM IUCV establishes a communication path between two z/VM guest
virtual machines on the same z/VM instance.
Because z/VM IUCV is independent from TCP/IP, Linux instances with no external
network can be accessed.

\fIterminal_id\fP identifies the z/VM IUCV connection and is similar to a port
number in TCP/IP communications.  \fIterminal_id\fP is case-sensitive and
consists of up to 8 alphanumeric characters.
It must be specified as a parameter in connection requests against
a \*t instance.

When a connection is established, \*t starts a login program.
When the login program ends, \*t also exits. Consider an
.BR inittab (5)
entry
.\"or an upstart job file
to ensure that \*t is restarted and ready for the next terminal session.
.
.
.
.SH OPTIONS
.TP
.BR \-\^h ", " \-\^\-help
Display a short help text and exit.
.
.TP
.BR \-\^v ", " \-\^\-version
Display the version number of \*t, then exit.
.
.TP
.BR \-\^a ", " \-\^\-allow-from " " \fIregex\fP
Limit permissions for incoming connections to z/VM user IDs that match
the regular expression \fIregex\fP. The connection is refused if the ID
does not match. If this parameter is omitted, connections are permitted
from any z/VM user ID.
.
.TP
.I login_program
\fIlogin_program\fP specifies the absolute path to the login program to be
started when a connection is established.
The default is
.BR /bin/login (1).
.br
The \fIlogin_program\fP must be separated by two dashes (--).
.
.TP
.I login_options
Depending on the particular login program used, \fIlogin_options\fP specifies
additional options.
.
.
.
.SH EXAMPLES
To allow remote logins using the terminal identifier "lnxterm":
.PP
.ft CR
.in +0.25in
.nf
iucvtty lnxterm
.fi
.in -0.25in
.ft

To only allow users from LNXSYS01 to connect to terminal "lnxterm":
.PP
.ft CR
.in +0.25in
.nf
iucvtty -a LNXSYS01 lnxterm
.fi
.in -0.25in
.ft

To only allow users from LNXSYS10 through LNXSYS19 to connect to terminal
"lnxterm":
.PP
.ft CR
.in +0.25in
.nf
iucvtty -a "LNXSYS1[0-9]" lnxterm
.fi
.in -0.25in
.ft

To use \fB/sbin/sulogin\fP instead of \fB/bin/login\fP for terminal "suterm":
.PP
.ft CR
.in +0.25in
.nf
iucvtty suterm -- /sbin/sulogin
.fi
.in -0.25in
.ft

An entry in \fB/etc/inittab\fP to facilitate user logins on terminal "lnxterm"
with \fB/bin/login\fP could be:
.PP
.ft CR
.in +0.25in
.nf
t1:2345:respawn:/usr/bin/iucvtty lnxterm
.fi
.in -0.25in
.ft

An entry in \fB/etc/inittab\fP to facilitate user logins on terminal "suterm"
with \fB/sbin/sulogin\fP in single user mode could be:
.PP
.ft CR
.in +0.25in
.nf
s1:S:respawn:/usr/bin/iucvtty suterm -- /sbin/sulogin
.fi
.in -0.25in
.ft
.
.
.
.SH DIAGNOSTICS
If the Linux kernel does not include kernel support for the AF_IUCV network
addressing family, \*t exits and displays the message
.I 'AF_IUCV address family is not available: Address family not supported by
.IR protocol' "."
.
.
.
.SH SECURITY
.SS Linux
The \*t program can be used by regular users.
Depending on the particular login program, \*t must be started with superuser
privileges for user authentication and authorization. For instance,
\fB/bin/login\fP requires superuser privileges and, thus, regular users
must use a different login program.

Each connection attempt is logged to the \fBauthpriv\fP syslog facility.

\*t uses pseudo-terminal (pts) devices to communicate with the
login program. For security reasons, some login programs, like
\fB/bin/login\fP, do not permit root logins on pseudo-terminal devices (see also
.BR /etc/securetty "(5))."

To permit root logins, consider using HVC terminal devices that are provided by
the z/VM IUCV hypervisor console (HVC) device driver.
.\"Enabling root logins on pseudo-terminal devices can compromise system security.
.\"To avoid this potential security exposure, consider using HVC terminal devices
.\"that are provided by the z/VM IUCV hypervisor console (HVC) device driver.
.
.
.
.SS z/VM guest virtual machine
See the
.BR af_iucv (7)
manual page for details about IUCV authorizations.
.
.
.
.SH "SEE ALSO"
.BR iucvconn (1),
.BR login (1),
.BR pts (4),
.BR regex (7),
.BR securetty (5),
.BR af_iucv (7),
.BR hvc_iucv (9)


CHANGELOG.md
Release history for s390-tools (MIT version)
--------------------------------------------

* __v2.xx.x (20xx-xx-xx)__

  For Linux kernel version: 6.x

  Add new tools / libraries:

  Changes of existing tools:

  Bug Fixes:

* __v2.38.0 (2025-06-25)__

  For Linux kernel version: 6.14 / 6.15

  Add new tools:
  - udev: New rule to set newly hotplugged CPUs online
  - zmemtopo: Display memory topology information
  - zpwr: Display power readings of a partition and CPC

  Removed tools / features:
  - check_hostkeydoc: Remove installation target
  - scsi_logging_level: Delete SCSI logging script (available in sg3_utils)
  - zdump: Drop build_arch for s390 DASD dumps
  - zdump: Drop non-extended multi-volume DASD dump support
  - zdump: Drop support of 32-bit dump architecture
  - zdump: Drop support of non-extended single volume DASD dumpers
  - zdump: Drop support of obsolete dumps and dumpers

  Changes of existing tools / libraries:
  - Various man-pages fixes
  - check_hostkeydoc: Add deprecation warning
  - check_hostkeydoc: Move to scripts directory
  - cpuplugd: Allow cpu hotplugging on systems without polarization
  - dbginfo.sh: Add Ubuntu snap tool
  - dbginfo.sh: Add missing config data and logs
  - dbginfo.sh: Reworking the container section
  - dbginfo.sh: Update for network commands
  - dbginfo.sh: Updating info for disks and lvm
  - libutil: Add machine type definition for machines 9175 and 9176
  - lscpumf: Add support for IBM z17 counter sets
  - lshwc: Add command line flag for run time
  - lshwc: Add flags to display counter values in hex
  - lshwc: Add output '--format' option
  - lshwc: Add support for delta counter value display
  - lspai: Add output '--format' option
  - lsreipl: Add secure boot state to output
  - lswhc: Add short names to lshwc output
  - pv_tools: Add Bash and Zsh completions
  - pvapconfig: Add '--unbind' option
  - pvimg/boot: Print error messages from stage3a bootloader
  - pvimg: Add support for CCK update
  - pvsecret: Add support for CCK update
  - pvsecret: Allow retrieving secrets by index & warn for duplicated entries
  - pvsecret: Deny adding secrets with duplicated secret IDs
  - zdev: Add support for virtio devices
  - zipl: Enhance mirror support
  - zipl: Implement '--dry-run' option for all dump jobs
  - zipl_helper.device-mapper: Support mirrors over NVMe devices
  - zkey/dracut: Add a dracut config file for zkey
  - zkey/initramfs: Update initramfs hook to correct drivers and include zkey plugins
  - zkey: Add support for converting a clear-key LUKS2 volume to use a secure key

  Bug Fixes:
  - chpstat: Add missing CMG 5 data fields
  - chpstat: Fix DPU utilization calculation
  - libutil/util_file: Handle over-read in util_file_read_fd()
  - pvattest: Fix successful 'check' evaluation
  - pvsecret: Fix some edge cases for plaintext keys
  - zipl_helper.device-mapper: Fix imprecise is_device_mapper() predicate
  - zkey: Fix EP11 secure key reencipher function
  - zpcictl: Fix command line parsing for invalid options

* __v2.37.0 (2025-02-07)__

  For Linux kernel version: 6.13

  Changes of existing tools:
  - dbginfo.sh: Add details on CPU-measurement
  - dbginfo.sh: Add new crypto command
  - dbginfo.sh: Add overview commands and crypto update
  - dbginfo.sh: Adding kdump info
  - dbginfo.sh: Removing outdated email references
  - dbginfo.sh: Rework network section
  - dbginfo.sh: Update copyright 2nd year
  - pvimg: Add '--(enable|disable)-image-encryption' flags to 'pvimg create'
  - pvimg: Add '--cck <FILE>' command line option and make '--comm-key' an alias
  - pvimg: Add '--hdr-key' command line option to 'pvimg create'
  - pvimg: Rename '--key' into '--hdr-key' and use '--key' as an alias (for 'pvimg info')
  - pvsecret: Add support for retrievable secrets
  - ziorep_config: Add PCHID field to adapter report
  - ziorep_traffic: Add DEVBUSID column to traffic report
  - ziorep_utilization: Add --fcp-device parameter to print virtual adapter report
  - ziorep_utilization: Add PCHID column to physical adapter report
  - ziorep_utilization: Now prints only physical adapter report by default
  - ziorep_utilization: Swap Bus-ID and CHPID columns in virtual adapter report
  - zipl/boot: Increase section size for eckd_mv dumper
  - zkey: Add support for listing and importing protected virtualization secrets

  Bug Fixes:
  - chpstat: Fix invalid utilization data on older kernels
  - opticsmon: Fix runaway loop in on_link_change()
  - zipl: Update inline assembly for GCC 15
  - zipl_helper.device-mapper: Add missed step in logical device resolution

* __v2.36.0 (2024-12-06)__

  For Linux kernel version: 6.12

  s390-tools: Define Rust MSRV as 1.75.0

  Add new tools / libraries:
  - cpacfinfo: Tool to provide CPACF information
  - opticsmon: Tools to monitor optical modules for directly attached PCI based NICs
  - pvimg: Rust rewrite of genprotimg

  Changes of existing tools:
  - chpstat: Add data bandwidth utilization column
  - chpstat: Add support for full CMCB
  - chpstat: Add support for new CMG types
  - dbginfo.sh: add overview commands and crypto update
  - genprotimg:
    1. genprotimg is now a symbolic link to the new tool `pvimg create`
    2. Breaking API changes in genprotimg:
       1. An existing output file is no longer silently overwritten; to revert
          to the old behavior use the `--overwrite` option.
       2. The Linux kernel component is now checked to determine whether it is a
          binary s390x kernel. To disable this behavior use the
          `--no-component-check` option.
  - hyptop: Support for structured output (json, json-seq, csv)
  - lszfcp: Add missing fallback marker for non-good fc_host port_state
  - lszfcp: Improve speed with many SCSI devices
  - pvattest: Add attestation policy check command
  - zipl: Add support of partitions of mirror md-devices

  Bug Fixes:
  - lszcrypt: Fix wrong state showing up for removed AP queue within SE guest

iucvterm/src/iucvtty.c
/*
 * iucvtty - Application that provides a full-screen terminal for iucvconn
 *
 * Core application
 *
 * Copyright IBM Corp. 2008, 2017
 *
 * s390-tools is free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See LICENSE for details.
 */
#include <errno.h>
#include <fcntl.h>
#include <pty.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/select.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <syslog.h>
#include <unistd.h>
#include <utmp.h>

#include "lib/util_base.h"

#include "af_iucv.h"
#include "iucvterm/config.h"
#include "iucvterm/functions.h"
#include "iucvterm/gettext.h"

#define SYSLOG_IDENT		"iucvtty"
#define PRG_COMPONENT           SYSLOG_IDENT
#define TERM_BUFSIZE		256
#define TERM_DEFAULT		"linux"


static volatile sig_atomic_t sig_shutdown;


/**
 * sig_handler() - Signal handler
 * @sig:	Signal number.
 */
static void sig_handler(int sig)
{
	sig_shutdown = sig;
}

/**
 * exec_login_prog() - execute a login program
 * @cmd:	Path to the (login) program executable
 */
static int exec_login_prog(char *cmd[])
{
	int rc;

	if (cmd != NULL)
		rc = execv(cmd[0], cmd);
	else
		rc = execl("/bin/login", "/bin/login", (char *) NULL);
	return rc;
}

/**
 * iucvtty_worker() - Handle an incoming client connection
 * @client:	Client file descriptor
 * @master:	PTY master file descriptor
 * @slave:	PTY slave file descriptor
 * @cfg:	IUCV TTY configuration structure.
 */
static int iucvtty_worker(int client, int master, int slave,
			  const struct iucvterm_cfg *cfg)
{
	int rc;
	struct iucvtty_msg *msg;
	pid_t child;
	fd_set set;
	size_t chunk;
	char term_env[TERM_BUFSIZE];


	/* flush pending terminal data */
	tcflush(master, TCIOFLUSH);

	/* read terminal parameters from client */
	if (iucvtty_rx_termenv(client, term_env, TERM_BUFSIZE))
		sprintf(term_env, TERM_DEFAULT);

	/* start login program */
	child = fork();
	if (child == -1) {
		print_error("Creating a new process to run the "
			    "login program failed");
		iucvtty_tx_error(client, ERR_FORK);
		return 1;	/* return from worker */
	}
	if (child == 0) {	/* child process */
		closelog();	/* close syslog */

		/* setup terminal */
		if (login_tty(slave)) {
			print_error("Setting up a terminal for user login failed");
			iucvtty_tx_error(client, ERR_SETUP_LOGIN_TTY);
			exit(2);
		}
		setenv("TERM", term_env, 1);
		if (exec_login_prog(cfg->cmd_parms)) {
			print_error("Running the login program failed");
			iucvtty_tx_error(client, ERR_CANNOT_EXEC_LOGIN);
		}
		exit(3);	/* we only reach here if exec has failed */
	}

	/* setup buffers */
	msg = malloc(MSG_BUFFER_SIZE);
	if (msg == NULL) {
		print_error("Allocating memory for the data buffer failed");
		rc = 2;
		goto out_kill_login;
	}

	/* multiplex i/o between login program and socket. */
	rc = 0;
	chunk = 0;
	while (!sig_shutdown) {
		FD_ZERO(&set);
		FD_SET(client, &set);
		FD_SET(master, &set);

		if (select(MAX(master, client) + 1, &set,
			   NULL, NULL, NULL) == -1) {
			if (errno == EINTR)
				continue;
			break;
		}

		if (FD_ISSET(client, &set)) {
			if (iucvtty_read_msg(client, msg,
					     MSG_BUFFER_SIZE, &chunk))
				break;
			switch (msg->type) {
			case MSG_TYPE_DATA:
				iucvtty_copy_data(master, msg);
				break;
			case MSG_TYPE_WINSIZE:
				if (msg->datalen != sizeof(struct winsize))
					break;
				if (ioctl(master, TIOCSWINSZ,
					  (struct winsize *) msg->data))
					print_error("Resizing the terminal "
						    "window failed");
				break;
			case MSG_TYPE_TERMIOS:	/* ignored */
				break;
			case MSG_TYPE_ERROR:
				iucvtty_error(msg);
				break;
			}
		}

		if (FD_ISSET(master, &set))
			if (iucvtty_tx_data(client, master,
					    msg, MSG_BUFFER_SIZE))
				break;
	}
	free(msg);

out_kill_login:
	/* ensure the chld is terminated before calling waitpid:
	 * - in case a sigterm has been received,
	 * - or a sigchld from other than the chld
	 */
	kill(child, SIGKILL);	/* cause a sigchld */
	waitpid(child, NULL, 0);

	return rc;
}

/**
 * main() - IUCV TTY program startup
 */
int main(int argc, char *argv[])
{
	struct iucvterm_cfg	conf;		/* program configuration */
	struct sockaddr_iucv	saddr, caddr;	/* IUCV socket address info */
	char 			client_host[9];	/* client guest name */
	int 			server, client;	/* socket file descriptors */
	int 			master, slave;	/* pre-allocated PTY fds */
	struct sigaction	sigact;		/* signal handler */
	int 			rc;
	socklen_t		len;


	/* gettext initialization */
	gettext_init();

	/* parse command line arguments */
	parse_options(PROG_IUCV_TTY, &conf, argc, argv);

	/* create server socket... */
	server = iucvtty_socket(&saddr, NULL, conf.service);
	if (server == -1) {
		print_error((errno == EAFNOSUPPORT)
			    ? N_("The AF_IUCV address family is not available")
			    : N_("Creating the AF_IUCV socket failed"));
		return 1;
	}
	if (bind(server, (struct sockaddr *) &saddr, sizeof(saddr)) == -1) {
		print_error("Binding the AF_IUCV socket failed");
		close(server);
		return 1;
	}
	if (listen(server, 1) == -1) {
		print_error("Listening for incoming connections failed");
		close(server);
		return 1;
	}

	/* pre-allocate PTY master/slave file descriptors */
	if (openpty(&master, &slave, NULL, NULL, NULL)) {
		print_error("Opening a new PTY master/slave device pair failed");
		close(server);
		return 1;
	}

	/* set close-on-exec for file descriptors */
	fcntl(master, F_SETFD, FD_CLOEXEC);
	fcntl(server, F_SETFD, FD_CLOEXEC);

	/* syslog */
	openlog(SYSLOG_IDENT, LOG_PID, LOG_AUTHPRIV);
	syslog(LOG_INFO, "Listening on terminal ID: %s, using pts device: %s",
		conf.service, ttyname(slave));

	rc = 0;
	len = sizeof(struct sockaddr_iucv);
	/* accept a new client connection */
	client = accept(server, (struct sockaddr *) &caddr, &len);
	if (client == -1) {
		print_error("An incoming connection could not be accepted");
		rc = 2;
		goto exit_on_error;
	}

	/* check if client is allowed to connect */
	userid_cpy(client_host, caddr.siucv_user_id);
	if (is_client_allowed(client_host, &conf)) {
		iucvtty_tx_error(client, ERR_NOT_AUTHORIZED);
		syslog(LOG_WARNING, "Rejected client connection from %s; "
				    "Client is not allowed to connect.",
				    client_host);
		rc = 3;

	} else { /* client is allowed to connect */
		syslog(LOG_INFO, "Accepted client connection from %s",
			client_host);
		/* set close-on-exec for client socket */
		fcntl(client, F_SETFD, FD_CLOEXEC);
		/* close server socket */
		close(server);

		/* setup signal handler to notify shutdown signal */
		sigemptyset(&sigact.sa_mask);
		sigact.sa_flags = SA_RESTART;
		sigact.sa_handler = sig_handler;
		if (sigaction(SIGCHLD, &sigact, NULL)
			|| sigaction(SIGTERM, &sigact, NULL)
			|| sigaction(SIGINT,  &sigact, NULL)
			|| sigaction(SIGPIPE, &sigact, NULL)) {
			print_error("Registering a signal handler failed");
			rc = 4;
			goto exit_on_error;
		}

		/* handle client terminal connection */
		rc = iucvtty_worker(client, master, slave, &conf);
	}

	close(client);

exit_on_error:
	close(slave);
	close(master);
	closelog();

	return rc;
}
